pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -0,0 +1,320 @@
+			if err != nil && !apierrors.IsNotFound(err) {
+				runtime.HandleError(fmt.Errorf("the IPAddress %s may have leaked, but can not be deleted: %v", ipAddress.Name, err))
+			}
+			continue
+		}
+
+		svc, err := r.serviceLister.Services(ipAddress.Spec.ParentRef.Namespace).Get(ipAddress.Spec.ParentRef.Name)
+		if apierrors.IsNotFound(err) {
+			// cleaning all IPAddress without an owner reference IF the time since it was created is greater than 60 seconds (default timeout value on the kube-apiserver)
+			// This is required because during the Service creation there is a time that the IPAddress object exists but the Service is still being creaated
+			// Assume that CreationTimestamp exist
+			ipLifetime := r.clock.Now().Sub(ipAddress.CreationTimestamp.Time)
+			if ipLifetime > 60*time.Second {
+				runtime.HandleError(fmt.Errorf("IPAddress %s appears to have leaked: cleaning up", ipAddress.Name))
+				r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressNotAllocated", "IPAddressAllocation", "IPAddress: %s for Service %s/%s appears to have leaked: cleaning up", ipAddress.Name, ipAddress.Spec.ParentRef.Namespace, ipAddress.Spec.ParentRef.Name)
+				err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+				if err != nil && !apierrors.IsNotFound(err) {
+					runtime.HandleError(fmt.Errorf("the IPAddress %s may have leaked, but can not be deleted: %v", ipAddress.Name, err))
+				}
+			}
+			continue
+		}
+		if err != nil {
+			r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "UnknownError", "IPAddressAllocation", "Unable to get parent Service for IPAddress %s due to an unknown error", ipAddress)
+			runtime.HandleError(fmt.Errorf("unable to get parent Service for IPAddress %s due to an unknown error: %v", ipAddress, err))
+			continue
+		}
+		// The service exists, we have checked in previous loop that all Service to IPAddress are correct
+		// but we also have to check the reverse, that the IPAddress to Service relation is correct
+		found := false
+		for _, clusterIP := range svc.Spec.ClusterIPs {
+			if ipAddress.Name == clusterIP {
+				found = true
+				break
+			}
+		}
+		if !found {
+			runtime.HandleError(fmt.Errorf("the IPAddress: %s for Service %s/%s has a wrong reference %#v; cleaning up", ipAddress.Name, svc.Name, svc.Namespace, ipAddress.Spec.ParentRef))
+			r.recorder.Eventf(ipAddress, nil, v1.EventTypeWarning, "IPAddressWrongReference", "IPAddressAllocation", "IPAddress: %s for Service %s/%s has a wrong reference; cleaning up", ipAddress.Name, svc.Namespace, svc.Name)
+			err := r.client.NetworkingV1alpha1().IPAddresses().Delete(context.Background(), ipAddress.Name, metav1.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				runtime.HandleError(fmt.Errorf("the IPAddress %s may have leaked, but can not be deleted: %v", ipAddress.Name, err))
+			}
+		}
+
+	}
+
+	return nil
```

Comment:
If you keep `ipaddress.kubernetes.io/ip-family` then you should probably check/repair that too.

Also, should we check that there are no `IPAddress` objects in the service CIDR ranges that _aren't_ managed by the service controller?